README.md

```diff
@@ -1,277 +1,899 @@
+resort to yield farming. These two-way swaps can go in either direction (eg, a two-way ADA <-> ERGO
+swap can go ADA -> ERGO or ERGO -> ADA). By allowing the liquidity provider to specify separate
+prices for each direction, they can profit from either direction of the swap. For example, if the
+market rate for DJED <-> USDC is 1:1, Alice can set DJED -> USDC to 1.01 USDC per DJED and USDC ->
+DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.
+
+Since the swaps are freely composable, arbitrarily complex swaps can be created. For example, Alice
+can chain ADA -> ERGO with ERGO -> DUST to create a transaction that converts ADA -> DUST. Not only
+does this composition allow for one-to-many conversions in a single transaction (eq, ADA to ERGO and
+DJED), but it also opens up the possibility for arbitragers to find profitable arbitrage
+opportunities. This potential for profit naturally incentivizes liquidity to be spread across all
+trading pairs instead of it being siloed into one trading pair as in TradFi and LP based DEXs.
+
+Finally, through the use of beacon tokens, the protocol is fully p2p. No batchers are required
+although they can be used if desired. And thanks to beacon tokens, users can get their own addresses
+which guarantees users maintain full delegation and voting control of their assets at all times.
+
+The only remaining challenge is the querying capacity of existing off-chain APIs, such as Blockfrost
+or Koios. (This is not a limitation for users with powerful enough hardware, as they can run their
+own API database).
+
+## Specification
 
-### Current DEX Deficiencies  
-One consequence of centralized script addresses is the necessity for liquidity pools and LP providers as discrete, permissioned entities. LPs are a common feature of many DEXes and have a number of downsides, namely slippage/impermanent-loss and lack of delegation control. Additionally,  current implementations of order-book style DEXes (which don't use LPs) suffer from the availability challenges of permissioned batchers. No matter how performant a system of batchers is, their resources do **not** scale in proportion to the number of users unless new batchers can join permissionlessly when demand is high. Lastly, permissioned protocols lack composability with other dApps; a critical feature for scaling eUTxO-based DeFi.
+Each type of swap is comprised of a validator / beacon policy pair. This means that each type of
+swap gets its own universal address. For example, all of Alice's one-way swaps, regardless of the
+trading pair, will be located at address X and all all of Alice's two-way swaps, regardless of the
+trading pair, will be located at address Y. Address X is != address Y. Furthermore, since each type
+of swap has its own beacon policy, each type is easily distinguishable off-chain while still being
+easily queryable by end users.
```

Comment:
Is this restriction necessary?  Why can't every swap be at a different address since the beacon tokens will still label the eUTxOs to make them easily searchable?

Reply:
It is not necessary but I thought managing a single address would be a better UX and make it easier for front-ends to integrate it. I don't want scenarios where users create a swap for an obscure trading pair and then forget about it. If it is in a different address, they may forget about it entirely. But if all swaps are in a single address, the user will see the swap the next time they check their DEX address. I personally don't know of any use cases that require separate addresses. If you really want a separate address, you can just use a separate staking credential.

Reply:
I see your point.  I thought there might be a way for your wallet + smart contract app to calculate additional addresses in a similar way to how wallets do this now by generating additional payment keys from one master key.

I am not sure of the pros and cons apart from my concern re quantum resistance and previously exposed public keys rather than just a hash of the public key.  But maybe this doesn't apply to smart contract public keys and hashes of public keys.  (See my other comment below related to this.)

I was mainly adding this comment to get you thinking because you obviously know more than me about such things.  I understand that quantum resistance is the main advantage wallets gain from generating additional addresses.